### PR TITLE
fix-pubsub-allrpc

### LINKED ISSUE #344: Examples use track.onremovetrack but should use stream.onremovetrack
### Your environment.
  - Version: macOS 10.15.7
  - Browser: Firefox 84.0
  - Other Information - fixed by changing to stream.onremovetrack and checking event for track kind before removing the video element (because stream.onremovetrack is called twice)

### What did you do?
Tested with pubusbtest, open two browser tabs, start publishing on both, close a browser tab and the remotes video element belonging to the closed browser tab is frozen and not removed.

### What did you expect?

The video element should be removed correctly when the other party leaves the page or closes the browser tab.

### What happened?

The video element remains on the page and shows a frozen picture.

### LINKED ISSUE #345: Pubsubtest example creates multiple video elements for the same stream in chrome
### Your environment.
  - Version: 10.15.7
  - Browser: Chrome 87.0.4280.88
  - Other Information

### What did you do?
Tested with pubsubtest example, used with multiple browsers to create several video streams. Wait a little, but happened as soon as 30 seconds one time.

### What did you expect?
Every stream should be associated with only a single video element.

### What happened?
The onumute event is fired multiple times in Chrome which causes creation of additional video elements.

### PR DESCRIPTION
**Pull request by adwpc**
#### Description
fix(pubsub html add multi player)
feat(allrpc add metrics)
fix(gallerytest html add muti player)
fix(Concurrency-grpc-stream-Send in different goroutines)
fix(open too many files error by setting ice.MulticastDNSModeDisabled)

#### Reference issue
Fixes #345 
Fixes #344

### CODE REVIEW

File: cmd/signal/grpc/server/server.go [hunk 1]
```diff
@@ -42,6 +43,7 @@ func NewServer(sfu *sfu.SFU) *SFUServer {
 // If the client closes this stream, the webrtc stream will be closed.
 func (s *SFUServer) Signal(stream pb.SFU_SignalServer) error {
 	peer := sfu.NewPeer(s.SFU)
+	var streamLock sync.Mutex
```
Comment (tarrencev) [inline cmd/signal/grpc/server/server.go:46]: why is this necessary?

File: pkg/sfu/sfu.go [hunk 2]
```diff
@@ -123,6 +124,7 @@ func NewWebRTCTransportConfig(c Config) WebRTCTransportConfig {
 	if len(c.WebRTC.Candidates.NAT1To1IPs) > 0 {
 		w.setting.SetNAT1To1IPs(c.WebRTC.Candidates.NAT1To1IPs, webrtc.ICECandidateTypeHost)
 	}
+	w.setting.SetICEMulticastDNSMode(ice.MulticastDNSModeDisabled)
```
Comment (tarrencev) [inline pkg/sfu/sfu.go:127]: should we make this configurable?

File: cmd/signal/grpc/server/server.go (hunk 1 above)
Comment (adwpc) [reply cmd/signal/grpc/server/server.go:46]: This is the error when i comment this lock:
`[2021-01-22 17:57:42.545] [ERROR] [132][peer.go][func4] => oniceconnectionstatechange error rpc error: code = Internal des
c = transport: transport: the stream is done or WriteHeader was already called`
The reason is:
https://github.com/grpc/grpc-go/blob/master/Documentation/concurrency.md#streams

File: pkg/sfu/sfu.go (hunk 2 above)
Comment (adwpc) [reply pkg/sfu/sfu.go:127]: Not required in most scenarios
Disable mdns can save many file descriptors

Comment (adwpc) [reply pkg/sfu/sfu.go:127]: Would like to make this configurable if someone need :)

File: cmd/signal/grpc/server/server.go (hunk 1 above)
Comment (tarrencev) [reply cmd/signal/grpc/server/server.go:46]: got it. i have seen that error. can we embed the lock in the sever instead?

```
type SFUServer struct {
	pb.UnimplementedSFUServer
        sync.Mutex
	SFU *sfu.SFU
}
```

Comment (adwpc) [reply cmd/signal/grpc/server/server.go:46]: Good idea. 
there is only one recv goroutine, so we only need a send lock.

File: pkg/sfu/sfu.go (hunk 2 above)
Comment (adwpc) [reply pkg/sfu/sfu.go]: made it configuabe

**Review by tarrencev — APPROVED** [review]

File: pkg/sfu/sfu.go (hunk 2 above)
Comment (OrlandoCo) [reply pkg/sfu/sfu.go]: @tarrencev @adwpc  There can be some security implication disabling the dns multicast for ice, so lets keep it enabled as default.

Comment (adwpc) [reply pkg/sfu/sfu.go]: OK

**Pull request merged**